feat: requestCollateral on /deposit

sangaman opened the pull request:
This adds an optional flag to the `/deposit` endpoint request to request collateral automatically after depositing.

I didn't want the deposit request to have to wait for the `client.requestCollateral` call to resolve, so it doesn't await that call but rather log to `this._logger.error` if it encounters an error and do nothing if it succeeds.

Related issue: https://github.com/ExchangeUnion/xud/issues/1756

erkarl commented on src/client.ts:
```diff
@@ -255,10 +255,14 @@ export default class Client {
+    });
+    
+    if (params.requestCollateral && params.assetId) {
+      client.requestCollateral(params.assetId).catch(this._logger.error);
```
Should we also wait for the collateral request to finish before returning the response?

sangaman replied:
I'm thinking it's better not to, partly based on the goal described here https://github.com/ExchangeUnion/xud/issues/1753 to make the deposit request return sooner - as soon as the channel is open. But @kilrau can weigh in if he thinks otherwise.

kilrau replied:
Hmm well we need to know if our collateral request was received and accepted by the node. So if the call returns directly after the node accepted our request (milliseconds), we wait for the call. If not, that behavior should probably be changed on connext side.

aryabhuptani replied:
Note that it's not possible to parallelize depositing/requesting collateral currently because of how sending funds into the channel works.

What's the verdict here? We're trying to move towards a pattern where we just try to resolve promises quickly and give you wrapped tx objects where you can separately just call `await tx.wait()` without blocking the whole process until tx gets mined (e.g. with your CLI).

If we're doing the above, then I wouldn't recommend making both the deposit/request collateral in a single call, as you would then be *forced* to wait for the deposit to be mined when making the deposit call.

Does thtat make sense?

kilrau replied:
So the flow would be `/deposit` and `await tx.wait()` as soon as that returns call `client.requestCollateral`. All without blocking the CLI. Correct?

sangaman replied:
@ArjunBhuptani Are you suggesting then to make a separate endpoint for collateral requests only? So we would call `/deposit` followed by `/request-collateral` or similar.

Also, I see how we could await a promise within this rest-api-client, but for the service communicating over http/rest (xud, in our case) we wouldn't have access to that promise directly. So if we wanted to act upon both the beginning of the tx and upon the tx completing, I figure we'd need an endpoint for the rest-client to send a request to upon transactions completing, right ?

But for the purposes here, if I understand correctly we would be fine just calling `/deposit`, completing the cli call, and then calling `/request-collateral` without the cli needing to wait for that to complete. I'll add more xud-specific thoughts in the related issue.

rhlsthrm replied:
@sangaman, we can definitely add the `request-collateral` method. This is already one of our top-level client methods, so it makes sense to expose it as a REST API method as well. We will make sure this is included in the next release.

> Also, I see how we could await a promise within this rest-api-client, but for the service communicating over http/rest (xud, in our case) we wouldn't have access to that promise directly. So if we wanted to act upon both the beginning of the tx and upon the tx completing, I figure we'd need an endpoint for the rest-client to send a request to upon transactions completing, right ?

One thing we could do is return the tx hash of the submitted tx which allows you to check the status on your own using an ETH provider.

sangaman replied:
Thanks. I'll put in a `request-collateral` endpoint then, seems like a sensible approach to me.

sangaman committed:
feat: /request-collateral endpoint

This adds a new endpoint that calls `client.requestCollateral` to
request collateral for a specified asset.

sangaman commented:
I'm closing this PR and opening a new one in its place since the approach of using a new endpoint is totally different from the approach here of calling it on `/deposit`.
